### PR TITLE
Support yarn 2

### PR DESCRIPTION
**Pull request by qianz2**
https://github.com/microsoft/Oryx/issues/895
https://github.com/Azure/static-web-apps/issues/287

### CODE REVIEW

File: src/Detector/Node/NodeDetector.cs [hunk 1]
```diff
@@ -122,9 +135,26 @@ public PlatformDetectorResult Detect(DetectorContext context)
...
             };
         }
 
+        private bool IsYarnLockFileYamlFile(ISourceRepo sourceRepo, string filePath)
+        {
+            var yamlContent = sourceRepo.ReadFile(filePath);
+            var yamlStream = new YamlStream();
+            try
+            {
+                yamlStream.Load(new StringReader(yamlContent));
```
Comment (gpcastro) [inline src/Detector/Node/NodeDetector.cs:149]: should we do any check to make sure it is indeed a yarn 2 file?

**Review by gpcastro — APPROVED** [review]

File: src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl [hunk 2]
```diff
@@ -15,11 +15,12 @@ echo
 # Yarn config is per user, and since the build might run with a non-root account, we make sure
 # the yarn cache is set on every build.
 YARN_CACHE_DIR=/usr/local/share/yarn-cache
+yarnCacheFolderName={{ YarnCacheFolderName }}
 if [ -d $YARN_CACHE_DIR ]
 then
 	echo
     echo "Configuring Yarn cache folder..."
-    yarn config set cache-folder $YARN_CACHE_DIR
+    yarn config set $yarnCacheFolderName $YARN_CACHE_DIR
```
Comment (arroyc) [inline src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl:23]: why is this change? are we passing the cache folder name to the script now?

File: src/BuildScriptGenerator/Node/NodeConstants.cs [hunk 3]
```diff
@@ -34,6 +34,9 @@ public static class NodeConstants
         public const string YarnVersionCommand = "echo Using Yarn version: && yarn --version";
         public const string HugoVersionCommand = "echo Using Hugo version: && hugo version";
         public const string YarnPackageInstallCommand = "yarn install --prefer-offline";
+        public const string Yarn2PackageInstallCommand = "yarn plugin import workspace-tools && yarn workspaces focus --all";
+        public const string Yarn1ConfigFolderName = "cache-folder";
+        public const string Yarn2ConfigFolderName = "cacheFolder";
```
Comment (arroyc) [inline src/BuildScriptGenerator/Node/NodeConstants.cs:39]: is this someting yarn2 has recommened?

File: src/BuildScriptGenerator/Node/NodePlatform.cs [hunk 4]
```diff
@@ -181,9 +182,18 @@ public PlatformDetectorResult Detect(RepositoryContext context)
             if (ctx.SourceRepo.FileExists(NodeConstants.YarnLockFileName))
             {
                 packageManagerCmd = NodeConstants.YarnCommand;
-                packageInstallCommand = NodeConstants.YarnPackageInstallCommand;
                 configureYarnCache = true;
                 packageInstallerVersionCommand = NodeConstants.YarnVersionCommand;
+                // yarn.lock file is valid YAML in Yarn 2+ and yarnrc.yml file replaces yarnrc in Yarn 2+.
```
Comment (arroyc) [inline src/BuildScriptGenerator/Node/NodePlatform.cs:187]: so a repo that doesnt have .yarnrc.yml but has yaml formatted yarn.lock will be considered as yarn2 supported ...

do we commit .yarnrc.yml in the repo too? if not then we may not need to check for this

**Review by arroyc — CHANGES_REQUESTED** [review]
I would also like to have a tests to verify the yamlfileformat checker is working fine. So a test that can assert true if a correct yaml file is passed as input and assert false when a non-yaml file is passed as input

File: src/BuildScriptGenerator/Node/NodePlatform.cs (hunk 4 above)
Comment (qianz2) [reply src/BuildScriptGenerator/Node/NodePlatform.cs:187]: yarn.lock file can be parsed as valid YAML format since yarn 2. The .yarnc.yml file does not have to be in the repo, it's a configuration file for yarn, sometimes it may not be in the repo for yarn2. So I added both as indicators.

File: src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl (hunk 2 above)
Comment (qianz2) [reply src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl:23]: This is simply the CLI command difference between yarn 1 and yarn 2, so we need to detect and pass in different command parameter here.

File: src/BuildScriptGenerator/Node/NodeConstants.cs (hunk 3 above)
Comment (qianz2) [reply src/BuildScriptGenerator/Node/NodeConstants.cs:39]: https://next.yarnpkg.com/getting-started/migration Yes, on the bottom CLI command section, it migrated yarn install command into this command.

File: src/Detector/Node/NodeDetector.cs (hunk 1 above)
Comment (qianz2) [reply src/Detector/Node/NodeDetector.cs:149]: It's used to parse yarn.lock file, if it's a valid YAML then it'll not throw any exceptions. I added tests for this part to make sure it works good.

File: src/BuildScriptGenerator/Node/NodeConstants.cs (hunk 3 above)
Comment (arroyc) [reply src/BuildScriptGenerator/Node/NodeConstants.cs:39]: @qianz2  not about yarn install ... looking at yarn cache config folder name. looks like yarn1 is cache-folder while yarn2 is cacheFolder. I'm curious why is that?

**Review by arroyc — APPROVED** [review]
thanks for adding two tests where detector returning true for yarn2, can you add 1 negative test where it returns false for yarn2 checks?

**Comment by Itrel** [comment]
Is there a way for me to utilize the Oryx build from this branch to test deployment of Azure static websites using yarn 2? 

What changes are needed to my GitHub actions in order to consume the Oryx build from this branch?

Is there a way I can clone Oryx and run it all locally?

Would it be a useful test case for you for me to attempt this?

File: src/BuildScriptGenerator/Node/NodePlatform.cs [hunk 5]
```diff
@@ -181,9 +182,18 @@ public PlatformDetectorResult Detect(RepositoryContext context)
             if (ctx.SourceRepo.FileExists(NodeConstants.YarnLockFileName))
             {
                 packageManagerCmd = NodeConstants.YarnCommand;
-                packageInstallCommand = NodeConstants.YarnPackageInstallCommand;
                 configureYarnCache = true;
                 packageInstallerVersionCommand = NodeConstants.YarnVersionCommand;
+                // yarn.lock file is valid YAML in Yarn 2+ and .yarnrc.yml file replaces .yarnrc in Yarn 2+.
+                // Applying yarn 2 cache folder name and package install command.
+                if (nodePlatformDetectorResult.HasYarnrcYmlFile
+                    || nodePlatformDetectorResult.IsYarnLockFileValidYamlFormat) {
```
Comment (merceyz) [inline src/BuildScriptGenerator/Node/NodePlatform.cs:190]: An alternative to this check is to test if the content of `yarn.lock` matches this regex
```regexp
/^__metadata:$/m
```

File: src/BuildScriptGenerator/Node/NodeConstants.cs (hunk 3 above)
Comment (merceyz) [reply src/BuildScriptGenerator/Node/NodeConstants.cs:39]: Since you're passing the `--all` flag but not the `--production` flag as well it's the same as if you were to run `yarn install`

> looks like yarn1 is cache-folder while yarn2 is cacheFolder. I'm curious why is that?

We made the settings camelCase in v2
